CPE200Lab1/CPE200Lab1/CalculatorEngine.cs

```diff
@@ -0,0 +1,223 @@
+			{
+				return;
+			}
+			if (isAfterEqual)
+			{
+				resetAll();
+			}
+			if (display.Length is 8)
+			{
+				return;
+			}
+			if (!hasDot)
+			{
+				display += ".";
+				hasDot = true;
+			}
+		}
+
+		public void removeLastDigit()
+		{
+			if (display is "Error")
+			{
+				return;
+			}
+			if (isAfterEqual)
+			{
+				return;
+			}
+			if (!isAllowBack)
+			{
+				return;
+			}
+			if (display != "0")
+			{
+				string current = display;
+				char rightMost = current[current.Length - 1];
+				if (rightMost is '.')
+				{
+					hasDot = false;
+				}
+				display = current.Substring(0, current.Length - 1);
+				if (display is "" || display is "-")
+				{
+					display = "0";
+				}
+			}
+		}
+		public void amount()
+		{
+			if (display is "Error")
+			{
+				return;
+			}
+			if (isAfterEqual)
+			{
+				resetAll();
+			}
+			if (isAfterOperater)
+			{
+				display = "0";
+			}
+			if (display.Length is 8)
+			{
+				return;
+			}
+			isAllowBack = true;
+			if (display is "0")
+			{
+				display = "";
+			}
+			display += figure;
+			isAfterOperater = false;
+		}
+		public string calculate(string operate, string firstOperand, string secondOperand, int maxOutputSize = 8)
+		{
+			switch (operate)
+			{
+				case "+":
+					return (Convert.ToDouble(firstOperand) + Convert.ToDouble(secondOperand)).ToString();
+				case "-":
+					return (Convert.ToDouble(firstOperand) - Convert.ToDouble(secondOperand)).ToString();
+				case "X":
+					return (Convert.ToDouble(firstOperand) * Convert.ToDouble(secondOperand)).ToString();
+				case "÷":
+					// Not allow devide be zero
+					if (secondOperand != "0")
+					{
+						double result;
+						string[] parts;
+						int remainLength;
+
+						result = (Convert.ToDouble(firstOperand) / Convert.ToDouble(secondOperand));
+						// split between integer part and fractional part
+						parts = result.ToString().Split('.');
+						// if integer part length is already break max output, return error
+						if (parts[0].Length > maxOutputSize)
+						{
+							return "E";
+						}
+						// calculate remaining space for fractional part.
+						remainLength = maxOutputSize - parts[0].Length - 1;
+						// trim the fractional part gracefully. =
+						return result.ToString("N" + remainLength);
+					}
+					break;
+				case "%":
+					//your code here
+					break;
+			}
+			return "E";
+		}
+		 
+		public void mark()
+		{
+			if (display is "Error")
+			{
+				return;
+			}
+			if (isAfterEqual)
+			{
+				resetAll();
+			}
+			// already contain negative sign
+			if (display.Length is 8)
+			{
+				return;
+			}
+			if (display[0] is '-')
+			{
+				display = display.Substring(1, display.Length - 1);
+			}
+			else
+			{
+				display = "-" + display;
+			}
+		}
+
+		public void balance()
+		{
+			if (display is "Error")
+			{
+				return;
+			}
+			string secondOperand = display;
+			string result = calculate(operate, firstOperand, secondOperand);
+			if (result is "E" || result.Length > 8)
```

Comment:
another problem of having "display" inside engine. Here, you can see that you've two error states, "E" and "Error", first problem, what if you've to change the interface to Thai, you'll need to replace all "Error" with "ผิดพลาด" in engine, which is not good. Second, you need to keep mapping between "E" and "Error" all the time.